sys/linux/socket_netlink_route_sched.txt

```diff
@@ -76,11 +90,37 @@ type tca_kind_options_t[NAME, VALUES] {
 	TCA_OPTIONS	nlattr[TCA_OPTIONS, VALUES]
 }
 
+type tca_kind_comp_t[NAME, OPTION, VALUES] {
+	TCA_KIND	nlattr[TCA_KIND, string[NAME]]
+	TCA_OPTIONS	nlattr[TCA_OPTIONS, OPTION]
+	TCA_OPTIONS	nlattr[TCA_OPTIONS, VALUES]
+}
+
 # ------------------------------ tc qdisc ------------------------------
-# TODO: qdisc_kind left "choke", "clsact", "codel", "drr", "dsmark", "bfifo", "pfifo", "pfifo_head_drop", "pfifo_fast", "fq", "fq_codel", "gred", "hfsc", "hhf", "htb", "ingress", "mqprio", "multiq", "netem", "pie", "prio", "qfq", "red", "rr", "sfb", "sfq", "tbf"
 qdisc_kind_options [
-	q_cbq	tca_kind_options_t["cbq", q_cbq_options]
-	q_cbs	tca_kind_options_t["cbs", q_cbs_options]
+	q_cbq			tca_kind_options_t["cbq", q_cbq_options]
+	q_cbs			tca_kind_options_t["cbs", q_cbs_options]
+	q_choke			tca_kind_options_t["choke", q_choke_options]
+	q_codel			tca_kind_options_t["codel", q_codel_options]
+	q_dsmark		tca_kind_options_t["dsmark", q_dsmark_options]
+	q_bfifo			tca_kind_options_t["bfifo", int32]
+	q_pfifo			tca_kind_options_t["pfifo", int32]
+	q_pfifo_head_drop	tca_kind_options_t["pfifo_head_drop", int32]
+	q_fq			tca_kind_options_t["fq", q_fq_options]
+	q_fq_codel		tca_kind_options_t["fq_codel", q_fq_codel_options]
+	q_gred			tca_kind_options_t["gred", q_gred_options]
+	q_hfsc			tca_kind_options_t["hfsc", int16]
+	q_hhf			tca_kind_options_t["hhf", q_hhf_options]
+	q_htb			tca_kind_options_t["htb", q_htb_options]
+	q_mqprio		tca_kind_comp_t["mqprio", tc_mqprio_qopt, q_mqprio_options]
```

Comment:
With the current description we will generate 
TCA_KIND attribute followed by
TCA_OPTIONS attribute with tc_mqprio_qopt struct as value followed by
TCA_OPTIONS attribute again with one of q_mqprio_options options as value

If I am reading net/sched/sch_mqprio.c correctly, we actually need:

TCA_KIND attribute followed by
TCA_OPTIONS attribute with value which is tc_mqprio_qopt struct followed by 0 or more q_mqprio_options attributes.

I.e. size of TCA_OPTIONS attribute includes both tc_mqprio_qopt and all of nested q_mqprio_options attributes. If that's correct then we need something along the following lines:

```
	q_mqprio		tca_kind_options_t["mqprio", tc_mqprio_message]

tc_mqprio_message {
    qopt   tc_mqprio_qopt
    attrs   array[q_mqprio_options]
}
```